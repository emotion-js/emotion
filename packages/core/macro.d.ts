declare module '@emotion/core/macro' {
  export * from '@emotion/core'
}
