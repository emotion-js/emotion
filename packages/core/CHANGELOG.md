# @emotion/core

## 10.0.21

### Patch Changes

- [7855db4a](https://github.com/emotion-js/emotion/commit/7855db4ae379f212e7b972b9108419154e17ed45) [#1537](https://github.com/emotion-js/emotion/pull/1537) Thanks [@Andarist](https://github.com/Andarist)! - Add missing `/* #__PURE__ */` annotation to creation of EmotionCssPropInternal

## 10.0.20

### Patch Changes

- [38bb2b19](https://github.com/emotion-js/emotion/commit/38bb2b19d4ff1552116829e068664516d167a3f8) [#1530](https://github.com/emotion-js/emotion/pull/1530) Thanks [@Andarist](https://github.com/Andarist)! - Fix class not being applied when putting nil as css prop value on wrapper component

## 10.0.17

### Patch Changes

- [66cda641](https://github.com/emotion-js/emotion/commit/66cda64128631790b81e3c9df273a972358ea593) [#1478](https://github.com/emotion-js/emotion/pull/1478) Thanks [@Andarist](https://github.com/Andarist)! - Add warnings about using illegal escape sequences
  - [10514a86](https://github.com/emotion-js/emotion/commit/10514a8635dcaa55b85c7bff90e2a9e14d1ba61f) [#1482](https://github.com/emotion-js/emotion/pull/1482) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Remove react native bundles in favour of different isBrowser detection
  - [66cda641](https://github.com/emotion-js/emotion/commit/66cda64128631790b81e3c9df273a972358ea593) [#1478](https://github.com/emotion-js/emotion/pull/1478) Thanks [@Andarist](https://github.com/Andarist)! - Update Babel dependencies

## 10.0.16

### Patch Changes

- [47262b64](https://github.com/emotion-js/emotion/commit/47262b64) - Fix labels from stack traces in some cases

## 10.0.15

### Patch Changes

- [188dc0e7](https://github.com/emotion-js/emotion/commit/188dc0e7) [#1443](https://github.com/emotion-js/emotion/pull/1443) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Extract styles rather than inline them in compat mode with the Global component
  - [4a3b18a2](https://github.com/emotion-js/emotion/commit/4a3b18a2) [#1451](https://github.com/emotion-js/emotion/pull/1451) Thanks [@Andarist](https://github.com/Andarist)! - Fixed issue with auto-labelling crashing on \$ at runtime

## 10.0.14

### Patch Changes

- [c0eb604d](https://github.com/emotion-js/emotion/commit/c0eb604d) [#1419](https://github.com/emotion-js/emotion/pull/1419) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Update build tool
