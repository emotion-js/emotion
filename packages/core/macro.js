module.exports = require('babel-plugin-emotion').macros.core
