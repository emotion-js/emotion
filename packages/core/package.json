{
  "name": "@emotion/core",
  "version": "10.0.17",
  "main": "dist/core.cjs.js",
  "module": "dist/core.esm.js",
  "browser": {
    "./dist/core.cjs.js": "./dist/core.browser.cjs.js",
    "./dist/core.esm.js": "./dist/core.browser.esm.js"
  },
  "types": "types/index.d.ts",
  "files": [
    "src",
    "dist",
    "types"
  ],
  "author": "mitchellhamilton <mitchell@mitchellhamilton.me>",
  "license": "MIT",
  "scripts": {
    "test:typescript": "dtslint types"
  },
  "dependencies": {
    "@babel/runtime": "^7.5.5",
    "@emotion/cache": "^10.0.17",
    "@emotion/css": "^10.0.14",
    "@emotion/serialize": "^0.11.10",
    "@emotion/sheet": "0.9.3",
    "@emotion/utils": "0.11.2"
  },
  "peerDependencies": {
    "react": ">=16.3.0"
  },
  "devDependencies": {
    "@emotion/styled": "^10.0.17",
    "@types/react": "^16.8.20",
    "dtslint": "^0.3.0",
    "emotion": "^10.0.17",
    "emotion-server": "^10.0.17",
    "emotion-theming": "^10.0.17",
    "html-tag-names": "^1.1.2",
    "svg-tag-names": "^1.1.1"
  },
  "repository": "https://github.com/emotion-js/emotion/tree/master/packages/core",
  "publishConfig": {
    "access": "public"
  },
  "umd:main": "dist/core.umd.min.js",
  "preconstruct": {
    "umdName": "emotionCore"
  }
}