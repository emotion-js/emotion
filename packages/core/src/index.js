// @flow
export { withEmotionCache, CacheProvider, ThemeContext } from './context'
export { jsx } from './jsx'
export { Global } from './global'
export { keyframes } from './keyframes'
export { ClassNames } from './class-names'
export { default as css } from '@emotion/css'
