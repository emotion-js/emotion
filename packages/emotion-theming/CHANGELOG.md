# emotion-theming

## 10.0.27

### Patch Changes

- [`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968) [#1698](https://github.com/emotion-js/emotion/pull/1698) Thanks [@Andarist](https://github.com/Andarist)! - Add LICENSE file
- Updated dependencies [[`4c62ae9`](https://github.com/emotion-js/emotion/commit/4c62ae9447959d438928e1a26f76f1487983c968)]:
  - @emotion/core@10.0.27
  - @emotion/styled@10.0.27
  - @emotion/weak-memoize@0.2.5

## 10.0.19

### Patch Changes

- [ffc7c58c](https://github.com/emotion-js/emotion/commit/ffc7c58c13277c8f05861e2534fc2a85fff01b76) [#1509](https://github.com/emotion-js/emotion/pull/1509) Thanks [@XeeD](https://github.com/XeeD)! - Add TypeScript type definition for the useTheme hook in emotion-theming

- Updated dependencies [c81c0033]:
  - @emotion/weak-memoize@0.2.4

## 10.0.18

### Patch Changes

- [bfb40050](https://github.com/emotion-js/emotion/commit/bfb400503d0d8d399e4e0051dc6e5eac40624b10) [#1499](https://github.com/emotion-js/emotion/pull/1499) Thanks [@tkh44](https://github.com/tkh44)! - Add useTheme React hook to emotion-theming

## 10.0.17

### Patch Changes

- [66cda641](https://github.com/emotion-js/emotion/commit/66cda64128631790b81e3c9df273a972358ea593) [#1478](https://github.com/emotion-js/emotion/pull/1478) Thanks [@Andarist](https://github.com/Andarist)! - Update Babel dependencies

## 10.0.14

### Patch Changes

- [c0eb604d](https://github.com/emotion-js/emotion/commit/c0eb604d) [#1419](https://github.com/emotion-js/emotion/pull/1419) Thanks [@mitchellhamilton](https://github.com/mitchellhamilton)! - Update build tool
  - [c673e200](https://github.com/emotion-js/emotion/commit/c673e200) [#1405](https://github.com/emotion-js/emotion/pull/1405) Thanks [@ryanswanson](https://github.com/ryanswanson)! - PropsOf<C> TypeScript utilities - Support defaultProps by delegating to new JSX and React types.
  - [33803731](https://github.com/emotion-js/emotion/commit/33803731) [#1397](https://github.com/emotion-js/emotion/pull/1397) Thanks [@SimeonC](https://github.com/SimeonC)! - Fixing flow types for the withTheme HOC
