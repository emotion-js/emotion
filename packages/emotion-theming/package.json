{
  "name": "emotion-theming",
  "version": "10.0.19",
  "description": "A CSS-in-JS theming solution, inspired by styled-components",
  "main": "dist/emotion-theming.cjs.js",
  "module": "dist/emotion-theming.esm.js",
  "types": "types/index.d.ts",
  "files": [
    "src",
    "dist",
    "types"
  ],
  "scripts": {
    "test:typescript": "dtslint types"
  },
  "repository": "https://github.com/emotion-js/emotion/tree/master/packages/emotion-theming",
  "keywords": [
    "react",
    "theme",
    "theming",
    "emotion",
    "cssinjs",
    "css-in-js"
  ],
  "author": "styled-components team",
  "contributors": [
    "Evan Scott <probablyup@gmail.com> (https://github.com/probablyup)"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/emotion-js/emotion/issues"
  },
  "homepage": "https://emotion.sh",
  "devDependencies": {
    "@emotion/core": "^10.0.17",
    "@emotion/styled": "^10.0.17",
    "@types/react": "^16.8.20",
    "dtslint": "^0.3.0",
    "react": "^16.5.2"
  },
  "dependencies": {
    "@babel/runtime": "^7.5.5",
    "@emotion/weak-memoize": "0.2.4",
    "hoist-non-react-statics": "^3.3.0"
  },
  "peerDependencies": {
    "@emotion/core": "^10.0.17",
    "react": ">=16.3.0"
  },
  "umd:main": "dist/emotion-theming.umd.min.js",
  "browser": {
    "./dist/emotion-theming.cjs.js": "./dist/emotion-theming.browser.cjs.js",
    "./dist/emotion-theming.esm.js": "./dist/emotion-theming.browser.esm.js"
  },
  "preconstruct": {
    "umdName": "emotionTheming"
  }
}