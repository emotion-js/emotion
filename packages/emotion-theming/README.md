# emotion-theming

> A CSS-in-JS theming solution for React

_`emotion-theming` is a theming library inspired by [styled-components](https://github.com/styled-components/styled-components)_

## Table of Contents

- [Install](#install)
- [Usage](#usage)
- [API](#api)
  - [ThemeProvider](#themeprovider-reactcomponenttype)
  - [withTheme](#withthemecomponent-reactcomponenttype-reactcomponenttype)
- [Credits](#credits)
- [License](#license)

## Install

```bash
# add --save if using npm@4 or lower
npm i emotion-theming

# or
yarn add emotion-theming
```

## Usage

Theming is accomplished by placing the `ThemeProvider` component, at the top of the React component tree and wrapping descendants with the `withTheme` higher-order component. This HOC gets the current theme and injects it as a "prop" into your own component.

The theme prop is automatically injected into components created with `styled`. The theme can also be accessed via passing a function to the css prop.

```jsx
// Page.js
import * as React from 'react'
/** @jsx jsx */
import { jsx } from '@emotion/core'
import styled from '@emotion/styled'

const Container = styled.div({
  background: 'whitesmoke',
  height: '100vh'
})

const Headline = styled.h1`
  color: ${props => props.theme.color};
  font-family: sans-serif;
`

export default class Page extends React.Component {
  render() {
    return (
      <Container>
        <Headline>I'm red!</Headline>
        <p css={theme => ({ color: theme.color })}>I'm also red!</p>
      </Container>
    )
  }
}

// App.js
import React from 'react'
import ReactDOM from 'react-dom'
import { ThemeProvider } from 'emotion-theming'

import Page from './Page.js'

const theme = {
  color: 'red'
}

class App extends React.Component {
  render() {
    return (
      <ThemeProvider theme={theme}>
        <Page />
      </ThemeProvider>
    )
  }
}
```

## API

### ThemeProvider: React.ComponentType

A React component that passes the theme object down the component tree via [context](https://reactjs.org/docs/context.html). Additional `ThemeProvider` components can be added deeper in the tree to override the original theme. The theme object will be merged into its ancestor as if by [`Object.assign`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign). If a function is passed instead of an object it will be called with the ancestor theme and the result will be the new theme.

_Accepts:_

- **`children`: React.Node**
- **`theme`: Object|Object => Object** - An object or function that provides an object.

```jsx
import React from 'react'
import styled from '@emotion/styled'
import { ThemeProvider, withTheme } from 'emotion-theming'

// object-style theme

const theme = {
  backgroundColor: 'green',
  color: 'red'
}

// function-style theme; note that if multiple <ThemeProvider> are used,
// the parent theme will be passed as a function argument

const adjustedTheme = ancestorTheme => ({ ...ancestorTheme, color: 'blue' })

class Container extends React.Component {
  render() {
    return (
      <ThemeProvider theme={theme}>
        <ThemeProvider theme={adjustedTheme}>
          <Text>Boom shaka laka!</Text>
        </ThemeProvider>
      </ThemeProvider>
    )
  }
}
```

> Note:
>
> Make sure to hoist your theme out of render otherwise you may have performance problems.

### withTheme(component: React.ComponentType): React.ComponentType

A higher-order component that provides the current theme as a prop to the wrapped child and listens for changes. If the theme is updated, the child component will be re-rendered accordingly.

```jsx
import PropTypes from 'prop-types'
import React from 'react'
import { withTheme } from 'emotion-theming'

class TellMeTheColor extends React.Component {
  render() {
    return <div>The color is {this.props.theme.color}.</div>
  }
}

TellMeTheColor.propTypes = {
  theme: PropTypes.shape({
    color: PropTypes.string
  })
}

const TellMeTheColorWithTheme = withTheme(TellMeTheColor)
```

## Credits

Thanks goes to the [styled-components team](https://github.com/styled-components/styled-components) and [their contributors](https://github.com/styled-components/styled-components/graphs/contributors) who designed this API.

## License

MIT 2017-present
