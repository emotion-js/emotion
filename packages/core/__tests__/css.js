// @flow
/** @jsx jsx */
import 'test-utils/next-env'
import { jsx, css, CacheProvider } from '@emotion/core'
import { ThemeProvider } from 'emotion-theming'
import renderer from 'react-test-renderer'
import createCache from '@emotion/cache'

const SomeComponent = (props: { lol: true }) => (props.lol ? 'yes' : 'no')

// test to make sure flow prop errors work.
// should probably try to make it so that components that require className props
// and have the css prop passed to them don't have type errors
// $FlowFixMe
;<SomeComponent /> // eslint-disable-line no-unused-expressions

test('thing', () => {
  const tree = renderer.create(
    <div>
      <div css={{ display: 'flex' }}>something</div>
    </div>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('css call composition', () => {
  let first = css`
    color: hotpink;
  `
  let tree = renderer.create(<div css={css({ ':hover': first })} />)

  expect(tree.toJSON()).toMatchSnapshot()
})

test('theming with the css prop', () => {
  const tree = renderer.create(
    <ThemeProvider theme={{ primary: 'hotpink' }}>
      <div css={theme => ({ color: theme.primary })} />
    </ThemeProvider>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('object with false', () => {
  const tree = renderer.create(
    <div>
      <div css={{ color: 'hotpink', display: false }}>something</div>
    </div>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('label in css call', () => {
  const tree = renderer.create(
    <div>
      <div
        css={css`
          color: hotpink;
          label: this-is-hotpink;
        `}
      >
        something
      </div>
    </div>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('string as css prop throws', () => {
  expect(() => {
    renderer.create(
      <div>
        <div
          css={`
            color: hotpink;
          `}
        >
          something
        </div>
      </div>
    )
  }).toThrowErrorMatchingSnapshot()
})

test('array fallback', () => {
  const tree = renderer.create(
    <div>
      <div
        css={{
          display: ['green', 'hotpink']
        }}
      >
        something
      </div>
    </div>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('nested at rule', () => {
  const tree = renderer.create(
    <div
      css={{
        '@media (min-width: 980px)': {
          backgroundColor: 'blue',
          '@supports (width: 100vw)': {
            backgroundColor: 'red'
          }
        }
      }}
    >
      something
    </div>
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('can set speedy via custom cache', () => {
  let cache = createCache({ speedy: true })
  renderer.create(
    <CacheProvider value={cache}>
      <div
        css={{
          color: 'hotpink'
        }}
      >
        <span css={{ color: 'yellow' }}>wow</span>
        something
      </div>
    </CacheProvider>
  )
  expect(cache.sheet.tags).toHaveLength(1)
})

test('autoLabel without babel', () => {
  let SomeComp = props => {
    return (
      <div
        {...props}
        css={{
          color: 'hotpink'
        }}
      >
        something
      </div>
    )
  }
  const tree = renderer.create(<SomeComp />)

  expect(tree.toJSON().props.className.endsWith('-SomeComp')).toBe(true)
})

test('overwrite styles from parent', () => {
  let SomeComponent = (props: Object) => (
    <div
      css={{
        color: 'green',
        backgroundColor: 'yellow'
      }}
      {...props}
    />
  )
  const tree = renderer.create(
    <SomeComponent
      css={{
        color: 'hotpink'
      }}
    />
  )

  expect(tree.toJSON()).toMatchSnapshot()
})

test('child selector array', () => {
  const tree = renderer.create(
    <div
      css={{
        ':hover': [{ color: 'green' }, { backgroundColor: 'yellow' }]
      }}
    />
  )

  expect(tree.toJSON()).toMatchSnapshot()
})
