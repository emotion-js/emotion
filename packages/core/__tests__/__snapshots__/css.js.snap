// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array fallback 1`] = `
.emotion-0 {
  display: green;
  display: hotpink;
}

<div>
  <div
    className="emotion-0"
  >
    something
  </div>
</div>
`;

exports[`child selector array 1`] = `
.emotion-0:hover {
  color: green;
  background-color: yellow;
}

<div
  className="emotion-0"
/>
`;

exports[`css call composition 1`] = `
.emotion-0:hover {
  color: hotpink;
}

<div
  className="emotion-0"
/>
`;

exports[`label in css call 1`] = `
.emotion-0 {
  color: hotpink;
}

<div>
  <div
    className="emotion-0"
  >
    something
  </div>
</div>
`;

exports[`nested at rule 1`] = `
@media (min-width:980px) {
  .emotion-0 {
    background-color: blue;
  }

@supports (width:100vw) {
    .emotion-0 {
      background-color: red;
    }
}
}

<div
  className="emotion-0"
>
  something
</div>
`;

exports[`object with false 1`] = `
.emotion-0 {
  color: hotpink;
}

<div>
  <div
    className="emotion-0"
  >
    something
  </div>
</div>
`;

exports[`overwrite styles from parent 1`] = `
.emotion-0 {
  color: green;
  background-color: yellow;
  color: hotpink;
}

<div
  className="emotion-0"
/>
`;

exports[`string as css prop throws 1`] = `
"Strings are not allowed as css prop values, please wrap it in a css template literal from '@emotion/css' like this: css\`
            color: hotpink;
          \`"
`;

exports[`theming with the css prop 1`] = `
.emotion-0 {
  color: hotpink;
}

<div
  className="emotion-0"
/>
`;

exports[`thing 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    className="emotion-0"
  >
    something
  </div>
</div>
`;
